# nextflow-ATACSeq

This workflow is based on the following nextflow pipes:

- https://github.com/mpg-age-bioinformatics/nf-fastqc
- https://github.com/mpg-age-bioinformatics/nf-flexbar
- https://github.com/mpg-age-bioinformatics/nf-mageck
- https://github.com/mpg-age-bioinformatics/nf-bowtie2
- https://github.com/mpg-age-bioinformatics/nf-macs
- https://github.com/mpg-age-bioinformatics/nf-diffbind
- https://github.com/mpg-age-bioinformatics/nf-CHIPseeker
- https://github.com/mpg-age-bioinformatics/nf-bedGraphToBigWig
- https://github.com/mpg-age-bioinformatics/nf-ATACseqQC

## local test

Create the test directory:
```
mkdir -p /tmp/nextflow_atac_local_complete_test
```

Download the demo data:
```
cd /tmp/nextflow_atac_local_complete_test
curl -J -O https://datashare.mpcdf.mpg.de/s/PcBd0kCPH7k82Z4/download
unzip raw_data.zip
```

Add the run script, paramaters file, and sample sheet to the project folder:
```
curl -J -O https://raw.githubusercontent.com/mpg-age-bioinformatics/nextflow-ATACSeq/main/params.local.json
curl -J -O https://raw.githubusercontent.com/mpg-age-bioinformatics/nextflow-ATACSeq/main/nextflow-ATACseq.local.sh
curl -J -O https://datashare.mpcdf.mpg.de/s/dUUKTLWHjYvNYh1/download
```

Run the local test:
```
bash nextflow-ATACseq.local.sh clone params.local.json
```

Most relevant results files will be found in `summary` eg.:
```

```

## flaski

If you have a json file generated by flaski.

```
LATEST_RELEASE=$(  curl --silent "https://api.github.com/repos/mpg-age-bioinformatics/nf-flaski-configs/releases/latest" | grep '"tag_name":' | sed -E 's/.*"([^"]+)".*/\1/' )
nextflow run mpg-age-bioinformatics/nf-flaski-configs -r ${LATEST_RELEASE} --raw /path/to/raw/data/folder --json /path/to/flaski/file.json --out /path/to/output/folder -profile local
curl -J -O https://raw.githubusercontent.com/mpg-age-bioinformatics/nextflow-rnaseq/main/nextflow-ATACSeq.local.sh
bash nextflow-ATACseq.local.sh clone /path/to/output/folder/params.json
```

